src/service/device.js

```diff
@@ -186,20 +198,66 @@ var Device = GObject.registerClass({
             lanBackend = this.service.manager.backends.get('lan');
 
         if (lanBackend && lanBackend.certificate)
-            localFingerprint = lanBackend.certificate.sha256();
+            localCert = lanBackend.certificate;
+
+
+        return [remoteCert, localCert];
+    }
+
+    // FIXME: backend should do this stuff
+    get encryption_info() {
+        let remoteFingerprint = _('Not available');
+        let localFingerprint = _('Not available');
+        const [remoteCert, localCert] = this.getCerts();
+
+        // Bluetooth connections have no certificate so we use the host address
+        if (this.connection_type === 'bluetooth') {
+            // TRANSLATORS: Bluetooth address for remote device
+            return _('Bluetooth device at %s').format('???');
+
+        // If the device is connected use the certificate from the connection
+        }
+
+        if (remoteCert)
+            remoteFingerprint = remoteCert.sha256();
+        if (localCert)
+            localFingerprint = localCert.sha256();
 
         // TRANSLATORS: Label for TLS Certificate fingerprint
         //
         // Example:
         //
         // Google Pixel Fingerprint:
-        // 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
+        // 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
         return _('%s Fingerprint:').format(this.name) + '\n' +
             remoteFingerprint + '\n\n' +
             _('%s Fingerprint:').format('GSConnect') + '\n' +
             localFingerprint;
     }
 
+    get verify_code() {
+        let verifyCode = _('Not available');
+        const [remoteCert, localCert] = this.getCerts();
+
+        // XXX MAKE THIS WORK
+        // let a = ByteArray.toGBytes(localCert.get_certificate());
+        // let b = ByteArray.toGBytes(remoteCert.certificate);
+        let a = ByteArray.toGBytes(ByteArray.fromString('a'));
+        let b = ByteArray.toGBytes(ByteArray.fromString('b'));
+        // Lexographic a < b comparison; this is what the < operator on QByteArray does
+        if (a.compare(b) < 0)
+            [a, b] = [b, a]; // swap
+
+
+        const checksum = new GLib.Checksum(GLib.ChecksumType.SHA256);
+        checksum.update(ByteArray.fromGBytes(a));
+        checksum.update(ByteArray.fromGBytes(b));
+        verifyCode = checksum.get_string();
+        checksum.free();
+
+        return verifyCode;
```

Comment:
Your comment mentioned segfaults... according to [the Checksum docs](https://gjs-docs.gnome.org/glib20~2.66.1/glib.checksum#method-get_string)...

> `get_string()`
> Returns:
> (`String`) — the hexadecimal representation of the checksum. The
> returned string is owned by the checksum and should not be modified
> or freed.

I'm gonna go out on a limb and say that means it's probably _NOT_ OK to access the Checksum-owned string _after_ freeing the Checksum. Perhaps, to be safe...?
```js
verifyCode = GLib.strdup(checksum.get_string());
```

Reply:
(Note that you'd then have to free the newly-allocated `verifyCode` String at some later point, once you're finished with it.)

Reply:
Well, that is bizarre so I've asked about it in the #javascript channel. SpiderMonkey should be able to handle this without the `GLib.strdup()`, but I'm surprise that function is even in our bindings and that it seems to work! `GLib.free()` is definitely a no-go in GJS, so I'm not sure if this will leak memory or if SM can free it.

Reply:
@andyholmes Heh, always fun when the soft underbelly of a native C API is exposed to a higher-level lanugage. Just like Python's lazy garbage collector frees developers from worrying much about allocation lifetimes... _in theory_. 

Until you find yourself working with SWIG-generated C/C++ library bindings in Python. Then, unless both the native library and its bindings are _really_ meticulous, you **definitely** start to care about your object lifetimes. In C++, not properly cleaning up your allocations "merely" leaks memory. But if Python's gc goes stomping around, cleaning up _**for**_ you, invalidating pointers it hasn't been told are still very much in use over on the native side, You're Gonna Have A Bad Time™.